include/deal.II/base/linear_index_iterator.h

```diff
@@ -0,0 +1,449 @@
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_linear_index_iterator_h
+#define dealii_linear_index_iterator_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+/**
+ * Many classes in deal.II, such as FullMatrix, TransposeTable, and
+ * SparseMatrix, are fundamentally views into contiguous buffers: put another
+ * way, one can iterate across each of these classes with a random access
+ * iterator. LinearIndexIterator provides most of the functionality needed to
+ * write iterators for these classes. LinearIndexIterator is essentially a
+ * simplified version of <code>boost::iterator_facade</code> that assumes
+ * <code>AccessorType</code> provides certain features (documented below). The
+ * main intent of this class is that containers can define their own accessor
+ * classes and then use the curiously recurring template pattern (CRTP)
+ * technique to define iterators, e.g.,
+ *
+ * @code
+ * template <typename T, bool Constness>
+ * class Accessor
+ * {
+ *   // const iterators store a const pointer
+ *   typedef typename std::conditional<Constness, const FullMatrix<T>*, FullMatrix<T>*>::type
+ *   object_pointer_type;
+ *
+ * protected:
+ *   // The design of LinearIndexIterator assumes that these two members exist
+ *   // and completely describe the iterator's state. linear_index should have a
+ *   // signed type.
+ *   std::ptrdiff_t linear_index;
+ *   object_pointer_type object;
+ *
+ *   // LinearIndexIterator needs access to linear_index and object.
+ *   friend class LinearIndexIterator<const_iterator<T>, Accessor<T, Constness>>;
+ *   friend class LinearIndexIterator<iterator<T>, Accessor<T, Constness>>;
+ * };
+ *
+ * class const_iterator : public LinearIndexIterator<const_iterator, Accessor<T, true>>
+ * {
+ *   // implement additional constructors here, but all state should be contained
+ *   // in the Accessor.
+ * };
+ *
+ * class iterator : public LinearIndexIterator<const_iterator, Accessor<T, false>>
+ * {
+ *   // implement additional constructors here, but all state should be contained
+ *   // in the Accessor.
+ * };
+ * @endcode
+ *
+ * The CRTP technique provides most of the boler-plate comparison operators
+ * for iterators. The implementation of this class assumes that
+ * <code>AccessorType</code> is a friend storing the offset
+ * <code>linear_index</code> and pointer <code>object</code>.
+ *
+ * @author David Wells, 2018
+ */
+template <class DerivedIterator, class AccessorType>
+class LinearIndexIterator
+{
+public:
+  /**
+   * Iterator category.
+   */
+  typedef std::random_access_iterator_tag iterator_category;
+
+  /**
+   * A typedef for the type you get when you dereference an iterator of the
+   * current kind.
+   */
+  typedef
+  AccessorType value_type;
+
+  /**
+   * Difference type.
+   */
+  typedef std::ptrdiff_t difference_type;
```

Comment:
You could probably generalize this if you wrote this as
```
  using difference_type = decltype(declval<Accessor>().linear_index - declval<Accessor>().linear_index);
```
i.e., the type of taking the difference between two `accessor.linear_index` numbers. This way, accessors don't need to use the large `ptrdiff_t` type if they don't want.

Not sure the obfuscation is worth the generalization, though.

Reply:
I like the generality, but we do iterator differences so infrequently (at least for the matrix classes) that I don't think the optimization is worth making things more complicated.